proposals/nnnn-safe-shared-contiguous-storage.md

```diff
@@ -0,0 +1,1197 @@
+
+#### Using `Span` with unsafe code:
+
+The `UnsafeBufferPointer` family of types can be be adapted for use with `Span`-taking API by using unsafe `Span`-providing functions. A `Span` instance obtained this way loses its static guarantee of temporal safety, because it is possible to deinitialize or deallocate the underlying memory before the end of the closure's scope. The closure provides a clear signal to avoid doing so, but it is not enforceable by the compiler.
+
+```swift
+extension UnsafeBufferPointer { // as well as UnsafeMutableBufferPointer
+  public func withUnsafeSpan<E: Error, Result: ~Copyable>(
+    _ body: (_ elements: Span<Element>) throws(E) -> Result
+  ) throws(E) -> Result
+
+  public func withUnsafeBytes<E: Error, Result: ~Copyable>(
+    _ body: (_ elements: RawSpan) throws(E) -> Result
+  ) throws(E) -> Result where Element: BitwiseCopyable
+}
+
+extension UnsafeRawBufferPointer { // as well as UnsafeMutableRawBufferPointer
+  public func withUnsafeBytes<E: Error, Result: ~Copyable>(
+    _ body: (_ elements: RawSpan) throws(E) -> Result
+  ) throws(E) -> Result
+}
+```
+
+`Span` and `RawSpan` also have an unsafe hatch for use with functions that require an unsafe argument, such as C functions:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  func withUnsafeBufferPointer<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeBufferPointer<Element>) throws(E) -> Result
+  ) throws(E) -> Result
+}
+
+extension Span where Element: BitwiseCopyable {
+  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
+  ) throws(E) -> Result
+}
+
+extension RawSpan {
+  func withUnsafeBytes<E: Error, Result: ~Copyable & ~Escapable>(
+    _ body: (_ buffer: UnsafeRawBufferPointer) throws(E) -> Result
+  ) throws(E) -> Result
+}
+```
+
+### Complete `Span` API:
+
+```swift
+public struct Span<Element: ~Copyable & ~Escapable>: Copyable, ~Escapable {
+  internal var _start: UnsafePointer<Element>
+  internal var _count: Int
+}
+```
+
+##### Basic API:
+
+The following properties, functions and subscripts have direct counterparts in the `Collection` protocol hierarchy. Their semantics shall be as described where they counterpart is declared (in `Collection` or `RandomAccessCollection`). The only difference with their counterpart should be a lifetime dependency annotation where applicable, allowing them to return borrowed nonescapable values or borrowed noncopyable values.
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  public var count: Int { get }
+  public var isEmpty: Bool { get }
+
+  public subscript(_ position: Int) -> Element { _read }
+}
+```
+
+##### Accessing subranges of elements:
```

Comment:
As I noted earlier on, all of the `extracting` functions depend on lifetime dependencies. Perhaps provide a closure-based `withSubspan`?